ETL/Pipeline/extract.py

```diff
@@ -0,0 +1,192 @@
+    if not isinstance(product_info, dict):
+        raise TypeError('product_info must be of type dict')
+
+    try:
+        discount = product_info["productPrice"]["discountPercentage"]
+        return discount > 0
+
+    except KeyError as e:
+        logging.error(f"Error getting sale status: {e}")
+        return None
+
+
+def process_product(product: dict) -> dict | None:
+    """Populates a single product dictionary with current price, reading time, and sale status."""
+    headers = {
+        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7)\
+        AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36'}
+
+    product_info = get_product_info(product, headers)
+
+    if product_info:
+        curr_price = get_current_price(product_info)
+        sale = get_sale_status(product_info)
+
+        if not sale is None and not curr_price is None:
+            product['current_price'] = curr_price
+            product['is_on_sale'] = sale
+            product['reading_at'] = datetime.now()
+            return product
+    logging.error(f"Error processing product {product['product_code']} ")
+    return None
+
+
+def extract_price_and_sales_data(product_list: list[dict]) -> list[dict]:
+    """Populates each product dictionary in the product list with current price, reading time,
+    and sale status using multiprocessing."""
+    logging.info("Starting Extraction")
+    with Pool(processes=4) as pool:
+        logging.info("Adding the current price and sale status")
+        results = list(pool.map(process_product, product_list))
+        logging.info("Finished Extraction")
+        logging.info("Removing erroneous / missing data")
+        results = [i for i in results if i is not None]
+    return results
+
+
+def has_required_keys(entry, required_keys):
+    """Check if all required keys are present in the dictionary."""
+    return all(key in entry for key in required_keys)
+
+
+def is_dict(entry):
+    """Check if the entry is a dictionary."""
+    return isinstance(entry, dict)
+
+
+def has_correct_types(entry, required_keys):
+    """Check if all required keys have the correct data type."""
+    return all(isinstance(entry[key], required_type) for key, required_type in required_keys.items())
+
+
+def convert_product_code(entry):
+    """Convert the product_code from str to int."""
+    try:
+        entry['product_code'] = int(entry['product_code'])
+        return True
+    except (ValueError, TypeError):
+        return False
+
+
+def validate_input(product_list):
+    """Validates the input list of products"""
+    required_keys = {
+        'product_id': int,
+        'url': str,
+        'product_code': int,
+        'product_name': str
+    }
+    valid_entries = []
+
+    for entry in product_list:
+        if is_dict(entry) and has_required_keys(entry, required_keys):
+            if convert_product_code(entry) and has_correct_types(entry, required_keys):
```

Comment:
Use `and`, there's not need to use nested ifs
```suggestion
        if (is_dict(entry) and has_required_keys(entry, required_keys) and
            convert_product_code(entry) and has_correct_types(entry, required_keys)):
```

Reply:
added!